### PR TITLE
Fixed #32468 -- Corrected usage of never_cache in contrib.admin. 

### PR DESCRIPTION
**Pull request by hakib**
…or on login and logout views in admin site.

Fixed [32468](https://code.djangoproject.com/ticket/32468#ticket)

### CODE REVIEW

**Comment by felixxm** [comment]
@hakib Thanks :+1: What do you think about adding a check to `never_cache()`?, e.g.
```python
if not isinstance(request, HttpRequest):
    raise TypeError(
        "never_cache didn't receive an HttpRequest. "
        "If you are decorating a classmethod, be sure to use "
        "@method_decorator."
    )
```

**Comment by hakib** [comment]
I'm afraid it might break code that took inspiration from Django admin... when I tried to override `login` on my custom admin site I used the decorator ad Django did, and I did not notice this issue until I tried to use another decorator that uses the request object.

Maybe a warning?

**Comment by felixxm** [comment]
> I'm afraid it might break code that took inspiration from Django admin... when I tried to override `login` on my custom admin site I used the decorator ad Django did, and I did not notice this issue until I tried to use another decorator that uses the request object.

Yes, but it will help to detect misuse of the decorator (like in your case), and fix is straightforward, so I would raise an error. I think it's acceptable.

**Comment by hakib** [comment]
Thanks @felixxm 

Please see the amended commit.

File: django/views/decorators/cache.py [hunk 1]
```diff
@@ -1,10 +1,24 @@
...
 
 
+def ensure_http_request(maybe_request):
+    """
+    Check that the argument is an HttpRequest, and fail
+    with a helpfull error if it's not.
+    """
+    if not isinstance(maybe_request, HttpRequest):
+        raise TypeError(
+            "never_cache didn't receive an HttpRequest. "
```
Comment (felixxm) [inline django/views/decorators/cache.py:16]: This is not a correct message for `@cache_control`, I'm going to move this check to decorators.

**Review by felixxm — COMMENTED** [review]
@hakib Thanks :+1: I pushed small edits.

**Comment by hakib** [comment]
Thanks @felixxm!

Let me know if there's anything else I can do.

**Pull request merged**